torch/onnx/_type_utils.py

```diff
@@ -108,36 +114,102 @@ def from_name(
     @classmethod
     @_beartype.beartype
     def from_dtype(cls, dtype: torch.dtype) -> JitScalarType:
-        """Convert a torch dtype to ScalarType."""
+        """Convert a torch dtype to JitScalarType.
+
+        Note: DO NOT USE this API when `dtype` comes from a `torch._C.Value.type()` calls.
+            A "RuntimeError: INTERNAL ASSERT FAILED at "../aten/src/ATen/core/jit_type_base.h" can
+            be raised in several scenarios where shape info is not present.
+            Instead use `from_value` API which is safer.
+
+        Args:
+            dtype: A torch.dtype to create a JitScalarType from
+
+        Returns:
+            JitScalarType
+
+        Raises:
+            ValueError: if dtype is not a valid torch.dtype or if it is None.
+        """
         if dtype not in _DTYPE_TO_SCALAR_TYPE:
             raise ValueError(f"Unknown dtype: {dtype}")
         return _DTYPE_TO_SCALAR_TYPE[dtype]
 
+    # TODO: `value: torch._C.Value` type annotation raises (CI machines only):
+    #       NameError: name '_C' is not defined
+    # In both cases, using typing.TypeVar or string literals didn't work
+    @classmethod
+    @_beartype.beartype
+    def from_value(cls, value, default=None) -> JitScalarType:
+        """Create a JitScalarType from a torch._C.Value's underlying scalar type.
```

Comment:
In summary of the several behavior change comments. I like that now we have an option of strict check with `default=None`, and will raise `SymbolicValueError` and fail early. Instead of returning `None` and causing problems later. We'll need to carefully ensure old code that expects a None does not start to raise after this change.

Reply:
all calls for `from_name` and `from_dtype` and hard-coded strings within `ScalarName = Literal[
    "Byte",
    "Char",
    "Double",
    "Float",
    "Half",
    "Int",
    "Long",
    "Short",
    "Bool",
    "ComplexHalf",
    "ComplexFloat",
    "ComplexDouble",
    "QInt8",
    "QUInt8",
    "QInt32",
    "BFloat16",
    "Undefined",
]` were refactored and are passing CI.